<a href="https://colab.research.google.com/github/dxvsh/LearningPytorch/blob/main/Week1/DLP_GA1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DLP GA1

In [1]:
from pprint import pprint

In [3]:
# install the datasets library
!pip install datasets > /dev/null

In [4]:
import datasets
from datasets import load_dataset, get_dataset_config_names, get_dataset_split_names, get_dataset_config_info

### **Q1.** (MSQ) Choose all the correct statements about the dataset at the following URL:

https://huggingface.co/datasets/ai4bharat/naamapadam

Naamapadam is a Named Entity Annotated dataset for 11 Indic languages.

**Note**: We strongly recommend using the appropriate functions in the datasets module to obtain the dataset's information, rather than accessing that information directly from the hub.

- [x] The dataset contains 11 configs(subsets)
- [ ] The dataset contains 22 configs(subsets)
- [x] The Hindi subset of the dataset contains 985787 training samples
- [ ] The Tamil subset of the dataset contains a total of 985787 samples.
- [x] The number of classes (labels,tags) is 7


In [6]:
configs = get_dataset_config_names('ai4bharat/naamapadam')
print(configs, len(configs))

['as', 'bn', 'gu', 'hi', 'kn', 'ml', 'mr', 'or', 'pa', 'ta', 'te'] 11


So, this dataset contains 11 configs/subsets

Using the `get_dataset_config_info()` method, we can get the info on a specific config/subset of a dataset

In [7]:
pprint(get_dataset_config_info('ai4bharat/naamapadam', 'hi').splits['train'])

SplitInfo(name='train',
          num_bytes=529397194,
          num_examples=985787,
          shard_lengths=[931000, 54787],
          dataset_name='naamapadam')


In [8]:
pprint(get_dataset_config_info('ai4bharat/naamapadam', 'ta').splits['train'])

SplitInfo(name='train',
          num_bytes=186464659,
          num_examples=497882,
          shard_lengths=None,
          dataset_name='naamapadam')


The hindi subset contains 985787 samples in the training split not the tamil one

In [9]:
pprint(get_dataset_config_info('ai4bharat/naamapadam', 'hi').features)

{'ner_tags': Sequence(feature=ClassLabel(names=['O',
                                                'B-PER',
                                                'I-PER',
                                                'B-ORG',
                                                'I-ORG',
                                                'B-LOC',
                                                'I-LOC'],
                                         id=None),
                      length=-1,
                      id=None),
 'tokens': Sequence(feature=Value(dtype='string', id=None), length=-1, id=None)}


In [8]:
# for config in configs:
#     pprint(get_dataset_config_info('ai4bharat/naamapadam', config).features)

If you run the above cell, you'll find that all the configs have the exact same features: `['O','B-PER','I-PER','B-ORG','I-ORG','B-LOC','I-LOC']`.

According to Gemini, the possible labels could mean:

- `O`: No special entity
- `B-PER`: Beginning of a person's name
- `I-PER`: Inside of a person's name
- `B-ORG`: Beginning of an organization name
- `I-ORG`: Inside of an organization name
- `B-LOC`: Beginning of a location name
- `I-LOC`: Inside of a location name

This is a common format for named entity recognition (NER) tasks.

So the number of classes is 7.

### **Q2.** (MSQ) Common data

Download the Tamil sub-dataset from the following URL https://huggingface.co/datasets/ai4bharat/naamapadam, and store in a variable named “ds”.

Find the location of the cache directory (one way to determine the cache directory location is by using a method from the Dataset class; please refer to the documentation). Which of the following files are in the cached directory corresponding to the downloaded dataset?


- [x] naamapadam-train.arrow
- [x] naamapadam-test.arrow
- [x] naamapadam-validation.arrow
- [ ] naamapadam-all.arrow
- [ ] dataset_info.json


In [11]:
# load the tamil subset

ds = load_dataset('ai4bharat/naamapadam', 'ta')

After a dataset has been loaded, you can use the `cache_files` property on the dataset object and it will return a list of cached files associated with that dataset

In [12]:
ds.cache_files

{'train': [{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-train.arrow'}],
 'test': [{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-test.arrow'}],
 'validation': [{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-validation.arrow'}]}

Another way of finding this:

The default datasets cache directory is:  `~/.cache/huggingface/datasets`, so you can look inside it to find the cached files of a particular dataset

In [13]:
!ls -lh ~/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc

total 180M
-rw-r--r-- 1 root root 1.4K Sep 27 10:30 dataset_info.json
-rw-r--r-- 1 root root   75 Sep 27 10:30 LICENSE
-rw-r--r-- 1 root root 314K Sep 27 10:30 naamapadam-test.arrow
-rw-r--r-- 1 root root 178M Sep 27 10:30 naamapadam-train.arrow
-rw-r--r-- 1 root root 1.0M Sep 27 10:30 naamapadam-validation.arrow


### **Q3.** What is the memory size of the downloaded dataset in MB?

From the above, we can see that the dataset size for the tamil set is about `180M`.

Let's also try to get this answer using python

In [14]:
ds.cache_files

{'train': [{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-train.arrow'}],
 'test': [{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-test.arrow'}],
 'validation': [{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-validation.arrow'}]}

In [15]:
import os

def get_total_dataset_size(dataset):
    cache_files = [dataset.cache_files['train'][0]['filename'],
                   dataset.cache_files['test'][0]['filename'],
                   dataset.cache_files['validation'][0]['filename']]

    total_size = sum(os.path.getsize(f) for f in cache_files)
    size_in_mb = total_size / (1024 * 1024)
    print(f"Total size: {size_in_mb:.2f} MB")

get_total_dataset_size(ds)

Total size: 179.29 MB


So, its about `180M` in size

### **Q4.** Enter the number of training samples in the dataset.

In [16]:
ds

DatasetDict({
    train: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 497882
    })
    test: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 758
    })
    validation: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 2795
    })
})

There are `497882` samples in the training split, `758` in test and `2795` in val.

In [17]:
# lets check a random sample to see what it looks like
ds['train'][30_000]

{'tokens': ['நிர்வாண',
  'புகைப்படங்களை',
  'இன்ஸ்டாகிராமில்',
  'வெளியிட்டார்',
  'இஷா',
  '.'],
 'ner_tags': [0, 0, 3, 0, 0, 0]}

**Q5.** Create a new column named “num tokens”. Compute the number of tokens (words) in each sample and store the results in the newly created column. Reassign the modified dataset to the same variable “ds”. How many tokens (in millions) are there in the entire dataset?
- [ ] 4
- [ ] 5
- [x] 6
- [ ] 7


In [18]:
def count_tokens(example):
    example['num_tokens'] = len(example['tokens'])
    return example

ds = ds.map(count_tokens)

Map:   0%|          | 0/497882 [00:00<?, ? examples/s]

Map:   0%|          | 0/758 [00:00<?, ? examples/s]

Map:   0%|          | 0/2795 [00:00<?, ? examples/s]

This `count_tokens` function that takes an example from the dataset and adds a new column num_tokens with the number of tokens in the tokens column. Then, it uses the map function to apply this function to each example in the dataset.

Let's check the dataset to make sure whether it really added a new column:

In [19]:
ds

DatasetDict({
    train: Dataset({
        features: ['tokens', 'ner_tags', 'num_tokens'],
        num_rows: 497882
    })
    test: Dataset({
        features: ['tokens', 'ner_tags', 'num_tokens'],
        num_rows: 758
    })
    validation: Dataset({
        features: ['tokens', 'ner_tags', 'num_tokens'],
        num_rows: 2795
    })
})

And looks like the new column was successfully added. Lets see a sample to make sure

In [20]:
ds['train'][30_000]

{'tokens': ['நிர்வாண',
  'புகைப்படங்களை',
  'இன்ஸ்டாகிராமில்',
  'வெளியிட்டார்',
  'இஷா',
  '.'],
 'ner_tags': [0, 0, 3, 0, 0, 0],
 'num_tokens': 6}

In [21]:
total_tokens = sum(ds['train']['num_tokens']) + sum(ds['test']['num_tokens']) + sum(ds['validation']['num_tokens'])
total_tokens_in_millions = total_tokens / 1_000_000
print(f"Total tokens: {total_tokens_in_millions:.2f} million")

Total tokens: 6.00 million


### **Q6.** The statement that the modified dataset increased the memory requirement about two fold is:

- [ ] True
- [x] False

If we again use `ls` to look at the cache directory, it seems that our modification created new cache files again but our memory requirement hasn't changed twofold. Its just that new cache files were created without removing the older ones that make it look like that

In [22]:
!ls -lh ~/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc

total 363M
-rw-r--r-- 1 root root 1.1M Sep 27 10:34 cache-1e5c1e90d25e5c48.arrow
-rw-r--r-- 1 root root 320K Sep 27 10:34 cache-7a0835b47199f1f1.arrow
-rw-r--r-- 1 root root 182M Sep 27 10:34 cache-bb530908781ec015.arrow
-rw-r--r-- 1 root root 1.4K Sep 27 10:30 dataset_info.json
-rw-r--r-- 1 root root   75 Sep 27 10:30 LICENSE
-rw-r--r-- 1 root root 314K Sep 27 10:30 naamapadam-test.arrow
-rw-r--r-- 1 root root 178M Sep 27 10:30 naamapadam-train.arrow
-rw-r--r-- 1 root root 1.0M Sep 27 10:30 naamapadam-validation.arrow


If we look at the `ds.cache_files` method, only the newly created cache files are the ones that are now associated with the dataset instead of all the files in the cache directory. The old ones don't contain our new modifications, they're just left there.

In [23]:
ds.cache_files

{'train': [{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/cache-bb530908781ec015.arrow'}],
 'test': [{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/cache-7a0835b47199f1f1.arrow'}],
 'validation': [{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/cache-1e5c1e90d25e5c48.arrow'}]}

The file size of these new files is pretty much the same as the old ones, there's not much of an increase, just a few MBs more

In [24]:
get_total_dataset_size(ds)

Total size: 183.14 MB


We can see that the size of the newly created files is just a little bit more than the old ones, it went from 179.2M to 183M

So, the memory increase of the dataset is not twofold, its just that the disk usage is increasing because the old files are also lying around

**Q7.** Concatenate all the samples across the splits in the following order: [train:test:validation]. Currently, each sample contains a list of words (tokens). Create a sentence by joining the individual words (tokens) in each sample using a single white space as a delimiter and store the resulting sample in a new column named “text”. Create a new dataset by removing the columns “ner tags” and “tokens”. Store the new dataset in the same variable “ds”. Enter the total number of samples in “ds”

In [25]:
# even if we don't do anything, the number of rows will still remain the same
ds['train'].num_rows + ds['test'].num_rows + ds['validation'].num_rows

501435

In [26]:
# Concatenate the splits
ds = datasets.concatenate_datasets([ds['train'], ds['test'], ds['validation']])

# Create the 'text' column
def create_text(example):
    example['text'] = ' '.join(example['tokens'])
    return example

ds = ds.map(create_text)

# Remove the 'ner_tags' and 'tokens' columns
ds = ds.remove_columns(['ner_tags', 'tokens'])

Map:   0%|          | 0/501435 [00:00<?, ? examples/s]

In [27]:
ds

Dataset({
    features: ['num_tokens', 'text'],
    num_rows: 501435
})

So there are `501435` total samples. Lets also check whether our modification applied correctly and if the new column was succesfully created.


In [28]:
ds[30_000]

{'num_tokens': 6,
 'text': 'நிர்வாண புகைப்படங்களை இன்ஸ்டாகிராமில் வெளியிட்டார் இஷா .'}

The text column now has a properly concatenated text of the tokens

**Q8.** Each modification to the dataset introduces new cache file in the cache directory, thereby increasing the disk space. The statement is
- [x] True
- [ ] False

In [29]:
!ls -lh ~/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc

total 660M
-rw-r--r-- 1 root root 297M Sep 27 10:36 cache-05ab4ef9eb274dcf.arrow
-rw-r--r-- 1 root root 1.1M Sep 27 10:34 cache-1e5c1e90d25e5c48.arrow
-rw-r--r-- 1 root root 320K Sep 27 10:34 cache-7a0835b47199f1f1.arrow
-rw-r--r-- 1 root root 182M Sep 27 10:34 cache-bb530908781ec015.arrow
-rw-r--r-- 1 root root 1.4K Sep 27 10:30 dataset_info.json
-rw-r--r-- 1 root root   75 Sep 27 10:30 LICENSE
-rw-r--r-- 1 root root 314K Sep 27 10:30 naamapadam-test.arrow
-rw-r--r-- 1 root root 178M Sep 27 10:30 naamapadam-train.arrow
-rw-r--r-- 1 root root 1.0M Sep 27 10:30 naamapadam-validation.arrow


In [30]:
ds.cache_files

[{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/cache-05ab4ef9eb274dcf.arrow'}]

So as you can see, again a new separate cache file was created for the concatenated dataset that we created and the older ones are still left, so any modification that we make introduces new cache files and increases the disk space.

**Q9.** Filter the dataset so that all the samples in the dataset should have at least six tokens (any symbol separated by a white space is considered a token). Enter the number of samples in the dataset after filtering. Enter the exact number. If the answer is 123456, enter it as 123456.

In [31]:
ds = ds.filter(lambda example: example['num_tokens']>=6)
ds

Filter:   0%|          | 0/501435 [00:00<?, ? examples/s]

Dataset({
    features: ['num_tokens', 'text'],
    num_rows: 370495
})

So there are `370495` samples which have atleast 6 tokens

**Q10.** Download all the splits of Tamil sub-dataset “inltkh.ta” of https://huggingface.co/datasets/ai4bharat/indic_glue .

Filter the dataset so that each sample contains at least six words (separated by a single white space).Then Interleave the resultant dataset with the above filtered dataset of naamapadam. Take 80% of samples from naamapadam and 20% from indic glue. Enter the number of samples after interleaving the datasets

Set the seed to 42 for interleaving.

In [32]:
# get_dataset_config_names('ai4bharat/indic_glue')

Download all splits of Tamil sub dataset

In [34]:
indic_glue_tamil = load_dataset('ai4bharat/indic_glue', 'inltkh.ta', split='train+test+validation')

In [35]:
indic_glue_tamil = indic_glue_tamil.filter(lambda example: len(example['text'].split())>=6)
indic_glue_tamil

Filter:   0%|          | 0/6684 [00:00<?, ? examples/s]

Dataset({
    features: ['text', 'label'],
    num_rows: 6416
})

In [36]:
from datasets import interleave_datasets

interleaved_ds = interleave_datasets([ds, indic_glue_tamil], probabilities=[0.8, 0.2], seed=42)
interleaved_ds

Dataset({
    features: ['num_tokens', 'text', 'label'],
    num_rows: 32290
})

There are `32290` samples after interleaving.